src/main.lua

```diff
@@ -993,40 +1023,64 @@ addEventingHandler(ActionMap.TokenCost, utils.hasMatchingTag("Action", ActionMap
 		if msg.Tags.Quantity then
 			assert(utils.isInteger(tonumber(msg.Tags.Quantity)), "Invalid quantity. Must be integer greater than 0")
 		end
+		assertValidFundFrom(fundFrom)
 	end
 
-	local inputStatus, inputResult = pcall(checkAssertions)
-
-	if not inputStatus then
+	local shouldContinue = eventingPcall(msg.ioEvent, function(error)
 		ao.send({
 			Target = msg.From,
 			Tags = { Action = "Invalid-Token-Cost-Notice", Error = "Bad-Input" },
-			Data = tostring(inputResult),
+			Data = tostring(error),
 		})
+	end, checkAssertions)
+	if not shouldContinue then
 		return
 	end
 
-	local status, result = pcall(arns.getTokenCost, {
-		intent = msg.Tags.Intent,
-		name = string.lower(msg.Tags.Name),
-		years = tonumber(msg.Tags.Years) or 1,
-		quantity = tonumber(msg.Tags.Quantity),
-		purchaseType = msg.Tags["Purchase-Type"] or "lease",
-		currentTimestamp = tonumber(msg.Timestamp) or tonumber(msg.Tags.Timestamp),
-	})
-	if not status then
+	local shouldContinue2, tokenCost = eventingPcall(
+		msg.ioEvent,
+		function(error)
+			ao.send({
+				Target = msg.From,
+				Tags = { Action = "Invalid-Token-Cost-Notice", Error = "Invalid-Token-Cost" },
+				Data = tostring(error),
+			})
+		end,
+		arns.getTokenCost,
+		{
+			intent = msg.Tags.Intent,
+			name = string.lower(msg.Tags.Name),
+			years = tonumber(msg.Tags.Years) or 1,
+			quantity = tonumber(msg.Tags.Quantity),
+			purchaseType = msg.Tags["Purchase-Type"] or "lease",
+			currentTimestamp = tonumber(msg.Timestamp) or tonumber(msg.Tags.Timestamp),
+		}
+	)
+	if not shouldContinue2 then
+		return
+	end
+
+	local shouldContinue3, fundingPlan = eventingPcall(msg.ioEvent, function(error)
 		ao.send({
 			Target = msg.From,
 			Tags = { Action = "Invalid-Token-Cost-Notice", Error = "Invalid-Token-Cost" },
-			Data = tostring(result),
-		})
-	else
-		ao.send({
-			Target = msg.From,
-			Tags = { Action = "Token-Cost-Notice", ["Token-Cost"] = tostring(result) },
-			Data = json.encode(result),
+			Data = tostring(error),
 		})
+	end, gar.getFundingPlan, msg.From, tokenCost, fundFrom)
+	if not shouldContinue3 then
+		return
 	end
+
+	ao.send({
+		Target = msg.From,
+		Tags = { Action = "Token-Cost-Notice", ["Token-Cost"] = tostring(tokenCost) },
+		Data = fundFrom and json.encode({
+				tokenCost = tokenCost,
+				fundingPlan = fundingPlan,
+			})
```

Comment:
mentioned in https://github.com/ar-io/ar-io-network-process/pull/172 - let's also include `discounts` as the result from `getTokenCost`